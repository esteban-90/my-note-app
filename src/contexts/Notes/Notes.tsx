import type { FC, PropsWithChildren } from 'react'
import type { Note, NoteUtils } from '@/types'
import { createContext, useContext } from 'react'
import useLocalStorage from 'use-local-storage'
import uniqid from 'uniqid'

const defaultValue: NoteUtils = {
  notes: [],
  getNote: () => undefined,
  addNote: () => undefined,
  updateNote: () => undefined,
  removeNote: () => undefined,
}

/**
 * Note context
 */

const context = createContext(defaultValue)

/**
 * Component for providing notes to children
 * @returns The provider
 */

export const NoteProvider: FC<PropsWithChildren> = ({ children }) => {
  const [notes, setNotes] = useLocalStorage('notes', defaultValue.notes)

  const getNote = (id: string) => {
    const note = notes.find((note) => note.id === id)
    return note
  }

  const addNote = (content: string) => {
    const newNote: Note = { id: uniqid('note-'), content, createdAt: new Date().toISOString() }
    const newNotes = notes.concat(newNote)
    setNotes(newNotes)
  }

  const updateNote = (id: string, content: string) => {
    const newNotes = notes.map((note) => (note.id === id ? { ...note, content } : note))
    setNotes(newNotes)
  }

  const removeNote = (id: string) => {
    const newNotes = notes.filter((note) => note.id !== id)
    setNotes(newNotes)
  }

  const { Provider } = context
  const value: NoteUtils = { notes, getNote, addNote, updateNote, removeNote }

  return <Provider value={value}>{children}</Provider>
}

/**
 * Custom hook for consumer components to use
 * @returns Utilities for working with notes
 */

export const useNotes = () => useContext(context)
