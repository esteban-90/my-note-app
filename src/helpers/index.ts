import Filter from 'bad-words-es'

/**
 * Checks if the content of a note has profane vocabulary.
 * @param noteContent The content of the note.
 * @returns true if it does, false otherwise.
 */
export const hasBadWords = (noteContent: string) => {
  const filter = new Filter()
  const isProfane = filter.isProfane(noteContent)
  return isProfane
}

/**
 * Retrieves the title of a note by splitting its content
 * until the first line break, or the first six words.
 * @param noteContent The content of the note.
 * @returns The resulting title.
 */
export const getTitle = (noteContent: string) => {
  let [title] = noteContent.split('\n')
  const splitted = title.split(' ')
  if (splitted.length > 6) title = splitted.slice(0, 6).join(' ')
  return title
}

/**
 * Formats the date of creation of a note according
 * to the current language locale set in the app.
 * @param noteCreation The date of creation, in ISO format.
 * @param locale The language locale.
 * @returns The formatted date.
 */
export const getDate = (noteCreation: string, locale: 'en' | 'es' = 'en') => {
  return new Date(noteCreation).toLocaleString(locale, { dateStyle: 'long', timeStyle: 'medium' })
}

/**
 * Retrieves the body of a note by removing its title and
 * getting the remaining words.
 * @param noteContent The content of the note.
 * @returns The resulting body.
 */
export const getBody = (noteContent: string) => {
  const title = getTitle(noteContent)
  const body = noteContent.replaceAll('\n', ' ').replaceAll(title, '')
  return body
}
