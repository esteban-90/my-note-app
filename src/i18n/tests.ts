import type { Resource } from 'i18next'
import { createInstance } from 'i18next'
import { initReactI18next } from 'react-i18next'
import { ns, supportedLngs } from './app'

const resources: Resource = {}
export const i18nInstance = createInstance()

void (async () => {
  for (const n of ns) {
    for (const lng of supportedLngs) {
      resources[lng] ??= {}
      resources[lng] = {
        ...resources[lng],
        [n]: await import(`../../public/translations/${lng}/${n}.json`),
      }
    }
  }
})()

i18nInstance.use(initReactI18next).init({
  ns,
  supportedLngs,
  fallbackLng: supportedLngs[0],
  lng: supportedLngs[0],
  react: { useSuspense: false },
  resources,
})
