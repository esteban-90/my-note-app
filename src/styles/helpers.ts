import { css } from '@emotion/react'

export const bgTransition = css`
  transition: background-color 0.7s;
`

export const centered = css`
  display: flex;
  align-items: center;
  justify-content: center;
`
