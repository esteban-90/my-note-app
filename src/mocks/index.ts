import type { Note, Language, Theme } from '@/types'
import uniqid from 'uniqid'
import { makeKey } from '@/helpers'

let seconds = 0

export const makeMockNote = (content: string): Note => {
  return {
    id: uniqid('note-'),
    content,
    createdAt: new Date(1995, 11, 17, 3, 24, seconds++).toISOString(),
  }
}

const makeMockNotes = (quantity = 1): Note[] => {
  return new Array(quantity).fill(null).map((_, index) => makeMockNote(`Hello test ${index + 1}\nTesting ${index + 1}`))
}

export const mockNotes = makeMockNotes(5)

export const mockNoteKey = makeKey('notes')
export const mockLanguageKey = makeKey('language')
export const mockThemeKey = makeKey('theme')

export const mockENLanguage: Language = 'en-US'
export const mockESLanguage: Language = 'es-PY'

export const mockDayTheme: Theme = 'day'
export const mockNightTheme: Theme = 'night'

export const formatMockDate = (date: string, locale: Language = 'en-US') => {
  return new Date(date).toLocaleString(locale, { dateStyle: 'long', timeStyle: 'medium' })
}
