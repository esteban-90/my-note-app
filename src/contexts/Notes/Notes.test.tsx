import { getTitle, getDate, getBody } from '@/helpers'
import { mockNotes, saveNotes, removeNotes } from '@/mocks'
import { screen, render, configure } from '@/tests'
import { NoteProvider, useNotes } from './Notes'

describe(NoteProvider.name, () => {
  let asFragment: () => DocumentFragment

  const Child = () => {
    const { notes } = useNotes()
    return (
      <ul>
        {notes.map(({ id, content, createdAt }) => (
          <li key={id}>
            <p>{getTitle(content)}</p>
            <p>{getBody(content)}</p>
            <p>{getDate(createdAt)}</p>
          </li>
        ))}
      </ul>
    )
  }

  const UI = (
    <NoteProvider>
      <Child />
    </NoteProvider>
  )

  beforeAll(() => {
    configure({ throwSuggestions: true })
    saveNotes()
  })

  beforeEach(() => {
    ;({ asFragment } = render(UI))
  })

  afterAll(() => {
    removeNotes()
  })

  it('should match snapshot', () => {
    expect(asFragment()).toMatchSnapshot()
  })

  it('should provide notes to child elements', () => {
    mockNotes.forEach(({ content, createdAt }) => {
      const noteTitle = getTitle(content)
      const noteTitleElement = screen.getByText(noteTitle)
      expect(noteTitleElement).toBeInTheDocument()

      const noteBody = getBody(content)
      const noteBodyElement = screen.getByText(noteBody, { trim: false })
      expect(noteBodyElement).toBeInTheDocument()

      const noteDate = getDate(createdAt)
      const noteDateElement = screen.getByText(noteDate)
      expect(noteDateElement).toBeInTheDocument()
    })
  })

  it('should provide the right number of notes', () => {
    const noteItems = screen.getAllByRole('listitem')
    expect(noteItems).toHaveLength(mockNotes.length)
  })
})
