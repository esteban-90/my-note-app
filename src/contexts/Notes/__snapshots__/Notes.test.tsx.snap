// Vitest Snapshot v1

exports[`NOTE PROVIDER: > should match snapshot 1`] = `
<DocumentFragment>
  <ul>
    <li>
      <p>
        Hello test 1
      </p>
      <p>
         Testing 1
      </p>
      <p>
        December 17, 1995 at 3:24:01 AM
      </p>
    </li>
    <li>
      <p>
        Hello test 2
      </p>
      <p>
         Testing 2
      </p>
      <p>
        December 17, 1995 at 3:24:02 AM
      </p>
    </li>
    <li>
      <p>
        Hello test 3
      </p>
      <p>
         Testing 3
      </p>
      <p>
        December 17, 1995 at 3:24:03 AM
      </p>
    </li>
    <li>
      <p>
        Hello test 4
      </p>
      <p>
         Testing 4
      </p>
      <p>
        December 17, 1995 at 3:24:04 AM
      </p>
    </li>
    <li>
      <p>
        Hello test 5
      </p>
      <p>
         Testing 5
      </p>
      <p>
        December 17, 1995 at 3:24:05 AM
      </p>
    </li>
  </ul>
</DocumentFragment>
`;
